### DISCUSSION TITLE
Flux compatibility with Mutating Webhooks

### DISCUSSION THREAD

**Discussion by velkovb** [fluxcd/flux2#2524] [Q&A] [answered]
Hello,

I am currently using Flux and planning to introduce some Mutating Webhooks in my setup to add nodeaffinity and tolerations to pods based on namespace or labels. Would something like that be compatible with Flux or will it detect the resource as changed after being mutated?

**Comment by stefanprodan** [comment, marked as answer]
It will work fine as long as those fields are not declared in Git.

**Reply by velkovb** [reply]
Thanks, and what will happen if those fields are present? I would guess if they are the same it would be fine but if they are different?

**Reply by stefanprodan** [reply]
In theory Flux should adopt the mutations as its own, but I haven't tested this, you'll have to do your own tests.

**Reply by velkovb** [reply]
Thanks, I will test and share the results.

**Reply by sthomson-wyn** [reply]
Any update on this @velkovb?

I'm currently evaluating argo vs flux, and there is an interesting difference in behaviour.

When [kyverno's autogen rules](https://kyverno.io/docs/writing-policies/autogen/) take place, argocd immediately reports the application as "out of sync", and attempts to reconcile the resources at the next sync. This makes sense, because the resources in the cluster are not the same as what is specified. There is also a means of [ignoring this diff](https://medium.com/@charled.breteche/using-kyverno-policies-with-argocd-1a600c1b249).

Flux, on the other hand, doesn't immediately report the difference. It appears to wait until the next sync interval, notice the difference, then reapply the manifest. Which is of course immediately changed by the mutation webhook. This means that flux is reapplying on _every single sync interval_. There also doesn't seem to be anything that I can find where you can tell flux to ignore the difference, like with argocd

To me, argocd's behaviour is much more in line with what I expect. The resource in the cluster _is_ different from what is specified, but flux reports it as "in sync", even though it is not.

**Reply by stefanprodan** [reply]
I guess the mutation webhook doesn't handle the dry-run, Flux detects changes by looking at the dry-run result and comparing it with the cluster state. If the webhook would mutate on dry-run (a thing that you can enable in its config I think), then Flux would work with it.

**Reply by velkovb** [reply]
@sthomson-wyn That probably comes also from the fact that argo renders the helm manifest and passes the output to kubectl and you don't have any helm releases there.

In our case 99% of the time the fields from the mutations were not in flux so we didn't have any issues.

**Reply by breygonsnow** [reply]
Just to ensure I am understanding Flux's determination of field changes, would a mutating webhook that introduces additional options in securityContext trigger Flux to redeploy.  And, if so, is there and option to ignore specific fields?

More specifically, I am testing the following and I believe that Flux is noticing this as a change:

Deployment yaml that Flux syncs, contains only the following, within the securityContext:

        securityContext:
          runAsUser: 48


We are leveraging Kyverno mutating webhooks to replace our old school PSPs to force some of the strict policies such as https://kyverno.io/policies/pod-security/restricted/restrict-seccomp-strict/restrict-seccomp-strict/ && https://kyverno.io/policies/pod-security/restricted/require-run-as-nonroot/require-run-as-nonroot/. 

The current policy, simply updates these fields (only relevant section copied for the Deployments, etc)

    mutate:
      foreach:
      - list: request.object.spec.template.spec.containers
        patchStrategicMerge:
          spec:
            template:
              spec:
                containers:
                - name: '{{ element.name }}'
                  securityContext:
                    allowPrivilegeEscalation: false
                    capabilities:
                      drop:
                      - ALL
                    runAsNonRoot: true
                    seccompProfile:
                      type: RuntimeDefault


Flux appears to discover this as a changed object and tries to redeploy it.


I0111 15:15:59.930634       1 mutation.go:94] WebhookServer "msg"="mutation rules from policy applied successfully" "action"="mutate" "gvk"="apps/v1, Kind=Deployment" "operation"="UPDATE" "resource”=“our_namespace/Deployment/oasis-attachments-dev" "policy"="change-privilege-escalation-match" "rules"=["change-privilege-escalation-match-controller-container"]

**Reply by steveizzle** [reply]
Any news or thoughts about this? Is this something you plan to look into? I find what @breygonsnow totally valid, as We bascially do the same thing (implementing kyverno mutating policies to set security best practices and offer the developers the possibility to define resources with flux). A constant reconciliation between two states does not sound like something desirable...

**Reply by stefanprodan** [reply]
Flux works with Kyverno mutations, the fact that Kyverno logs the dry run has no relevance, Flux performs dry runs on the interval set by users, but that doesn’t mean it detects conflicts.
